File: backend/src/routes/clubs.ts
import express from 'express'
import { prisma } from '../lib/prisma'
import { clubCreateSchema, clubIdSchema, clubUpdateSchema } from '../schemas/club'
import { ClubService } from '../services/ClubService'
import { createRequireAdmin } from '../middleware/adminAuth'
import { env } from '../lib/env'

const router = express.Router()
const clubService = new ClubService(prisma)
const requireAdmin = createRequireAdmin(env.SESSION_SECRET)

router.get('/', async (_req, res, next) => {
  try { res.json({ success: true, data: await clubService.getAllClubs() }) }
  catch (error) { next(error) }
})

router.get('/statistics/summary', async (_req, res, next) => {
  try { res.json({ success: true, data: await clubService.getStatistics() }) }
  catch (error) { next(error) }
})

router.get('/search/:keyword', async (req, res, next) => {
  try { res.json({ success: true, data: await clubService.searchClubs(req.params.keyword.slice(0, 100)) }) }
  catch (error) { next(error) }
})

router.get('/tags/all', async (_req, res, next) => {
  try { res.json({ success: true, data: await clubService.getAllTags() }) }
  catch (error) { next(error) }
})

router.get('/category/:category', async (req, res, next) => {
  try { res.json({ success: true, data: await clubService.filterByCategory(req.params.category.slice(0, 30)) }) }
  catch (error) { next(error) }
})

router.get('/:id', async (req, res, next) => {
  try {
    const club = await clubService.getClubById(clubIdSchema.parse(req.params.id))
    if (!club) return res.status(404).json({ success: false, error: 'NOT_FOUND', message: '社团不存在' })
    return res.json({ success: true, data: club })
  } catch (error) { return next(error) }
})

router.post('/', (req, res, next) => {
  try { res.locals.clubData = clubCreateSchema.parse(req.body); next() } catch (error) { next(error) }
}, requireAdmin, async (_req, res, next) => {
  try {
    const club = await clubService.createClub(res.locals.clubData)
    res.status(201).json({ success: true, data: club })
  } catch (error) { next(error) }
})

router.put('/:id', (req, res, next) => {
  try {
    res.locals.clubId = clubIdSchema.parse(req.params.id)
    res.locals.clubData = clubUpdateSchema.parse(req.body)
    next()
  } catch (error) { next(error) }
}, requireAdmin, async (_req, res, next) => {
  try {
    const club = await clubService.updateClub(res.locals.clubId, res.locals.clubData)
    res.json({ success: true, data: club })
  } catch (error) { next(error) }
})

router.delete('/:id', (req, res, next) => {
  try { res.locals.clubId = clubIdSchema.parse(req.params.id); next() } catch (error) { next(error) }
}, requireAdmin, async (_req, res, next) => {
  try {
    const club = await clubService.deleteClub(res.locals.clubId)
    res.json({ success: true, data: club })
  } catch (error) { next(error) }
})

export default router

File: backend/app/routes/clubs.py
"""社团 CRUD 与查询路由。等价于原 TS routes/clubs.ts。"""
from __future__ import annotations

from fastapi import APIRouter, Depends, Path

from ..deps import club_service
from ..errors import AppError, envelope
from ..schemas import ClubCreate, ClubUpdate
from ..security import require_admin
from ..serializers import club_json, clubs_json

router = APIRouter()


@router.get("")
async def list_clubs():
    return envelope(clubs_json(club_service.get_all_clubs()))


@router.get("/statistics/summary")
async def statistics_summary():
    return envelope(club_service.get_statistics())


@router.get("/search/{keyword}")
async def search_clubs(keyword: str):
    return envelope(clubs_json(club_service.search_clubs(keyword[:100])))


@router.get("/tags/all")
async def all_tags():
    return envelope(club_service.get_all_tags())


@router.get("/category/{category}")
async def by_category(category: str):
    return envelope(clubs_json(club_service.filter_by_category(category[:30])))


@router.get("/{club_id}")
async def get_club(club_id: int = Path(gt=0)):
    club = club_service.get_club_by_id(club_id)
    if club is None:
        raise AppError(404, "NOT_FOUND", "社团不存在")
    return envelope(club_json(club))


@router.post("", status_code=201, dependencies=[Depends(require_admin)])
async def create_club(payload: ClubCreate):
    club = club_service.create_club(payload.model_dump(by_alias=False))
    return envelope(club_json(club))


@router.put("/{club_id}", dependencies=[Depends(require_admin)])
async def update_club(payload: ClubUpdate, club_id: int = Path(gt=0)):
    data = payload.model_dump(by_alias=False, exclude_unset=True)
    club = club_service.update_club(club_id, data)
    return envelope(club_json(club))


@router.delete("/{club_id}", dependencies=[Depends(require_admin)])
async def delete_club(club_id: int = Path(gt=0)):
    club = club_service.delete_club(club_id)
    return envelope(club_json(club))
